packages/pds/src/sql-repo-storage.ts

```diff
@@ -0,0 +1,200 @@
+  }
+
+  async commitStaged(commit: CID, prev: CID | null): Promise<void> {
+    this.db.assertTransaction()
+    const commitBlocks: RepoCommitBlock[] = []
+    this.staged.forEach((_bytes, cid) => {
+      commitBlocks.push({
+        commit: commit.toString(),
+        block: cid.toString(),
+      })
+    })
+    const insertBlocks = this.putMany(this.staged)
+    const insertCommit = this.db.db
+      .insertInto('repo_commit_block')
+      .values(commitBlocks)
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    const updateRoot =
+      prev === null ? this.insertRoot(commit) : this.updateRoot(commit, prev)
+    const insertCommitHistory = this.db.db
+      .insertInto('repo_commit_history')
+      .values({
+        commit: commit.toString(),
+        prev: prev ? prev.toString() : null,
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    await Promise.all([
+      insertBlocks,
+      insertCommit,
+      updateRoot,
+      insertCommitHistory,
+    ])
+    this.clearStaged()
+  }
+
+  private async insertRoot(commit: CID): Promise<void> {
+    await this.db.db
+      .insertInto('repo_root')
+      .values({
+        did: this.did,
+        root: commit.toString(),
+        indexedAt: this.getTimestamp(),
+      })
+      .execute()
+  }
+
+  private async updateRoot(commit: CID, prev: CID): Promise<void> {
+    const res = await this.db.db
+      .updateTable('repo_root')
+      .set({
+        root: commit.toString(),
+        indexedAt: this.getTimestamp(),
+      })
+      .where('did', '=', this.did)
+      .where('root', '=', prev.toString())
+      .executeTakeFirst()
+    if (res.numUpdatedRows < 1) {
+      throw new Error('failed to update repo root: misordered')
+    }
+  }
+
+  private getTimestamp(): string {
+    return this.timestamp || new Date().toISOString()
+  }
+
+  async getCommitPath(
+    latest: CID,
+    earliest: CID | null,
+  ): Promise<CID[] | null> {
+    const res = await this.db.db
+      .withRecursive('ancestor(commit, prev)', (cte) =>
+        cte
+          .selectFrom('repo_commit_history as commit')
+          .select(['commit.commit as commit', 'commit.prev as prev'])
+          .where('commit', '=', latest.toString())
+          .unionAll(
+            cte
+              .selectFrom('repo_commit_history as commit')
+              .select(['commit.commit as commit', 'commit.prev as prev'])
+              .innerJoin('ancestor', 'ancestor.prev', 'commit.commit')
+              .if(earliest !== null, (qb) =>
+                // @ts-ignore
+                qb.where('commit.commit', '!=', earliest?.toString() as string),
+              ),
+          ),
+      )
+      .selectFrom('ancestor')
+      .select('commit')
+      .execute()
+    return res.map((row) => CID.parse(row.commit)).reverse()
+  }
+
+  async destroySaved(): Promise<void> {
+    throw new Error('Destruction of SQL repo storage not allowed at runtime')
+  }
```

Comment:
I think we may actually need this fairly soon.  I could still imagine it not living on this class, though.